includes/validation/class-amp-validation-manager.php

```diff
@@ -1585,11 +1586,52 @@ public static function get_source( $callback ) {
 
 		$source = compact( 'reflection' );
 
+		$file   = wp_normalize_path( $reflection->getFileName() );
+		$source = array_merge( $source, self::get_file_source( $file ) );
+
+		// If a file was identified, then also supply the line number.
+		if ( isset( $source['file'] ) ) {
+			$source['line'] = $reflection->getStartLine();
+		}
+
+		if ( $reflection instanceof ReflectionMethod ) {
+			$source['function'] = $reflection->getDeclaringClass()->getName() . '::' . $reflection->getName();
+		} else {
+			$source['function'] = $reflection->getName();
+		}
+
+		return $source;
+	}
+
+	/**
+	 * Identify the type, name, and relative path for a file.
+	 *
+	 * @since 2.0.1
+	 *
+	 * @param string $file File.
+	 * @return array {
+	 *     @type string $type Source type (core, plugin, mu-plugin, or theme).
+	 *     @type string $name Source name.
+	 *     @type string $file Relative file path based on the type.
+	 * }
+	 */
+	public static function get_file_source( $file ) {
+		if (
+			! isset( self::$template_directory )
+			||
+			! isset( self::$template_slug )
+			||
+			! isset( self::$stylesheet_directory )
+			||
+			! isset( self::$stylesheet_slug )
+		) {
```

Comment:
These conditions can be combined:
```suggestion
		if ( ! isset( self::$template_directory, self::$template_slug, self::$stylesheet_directory, self::$stylesheet_slug ) ) {
```

Reply:
Not quite, as that combined usage would be the same as:

```php
			! isset( self::$template_directory )
			&&
			! isset( self::$template_slug )
			&&
			! isset( self::$stylesheet_directory )
			&&
			! isset( self::$stylesheet_slug )
```

Reply:
Alternatively, we could eliminate this condition and invocation of `set_theme_variables()` here entirely, and instead call `AMP_Validation_Manager::set_theme_variables()` from inside of `amp_is_available()`. That seems somewhat better. Or `get_file_source()` could take a second `$update_theme_variables` arg which when set would call the method.

Reply:
From the [`isset` docs](https://www.php.net/manual/en/function.isset.php):

> If multiple parameters are supplied then isset() will return TRUE only if all of the parameters are considered set. Evaluation goes from left to right and stops as soon as an unset variable is encountered.

So it will indeed be the same as what you had.

Reply:
Maybe this is confusing because of the negation. Do note that in the combined `isset()` the entire construct is negated as a whole, not every individual check.

So, `isset( a, b, c )` will be **true** if all of a, b & c are set => will be **false** if **either a or b or c is not set**.
Therefore `! isset( a, b, c )` will be **false** if all of a, b & c are set => will be **true** if **either a or b or c is not set**.

Reply:
Or to paraphrase in the code you've used as the bad example above...
It's not this:
```php
			! isset( self::$template_directory )
			&&
			! isset( self::$template_slug )
			&&
			! isset( self::$stylesheet_directory )
			&&
			! isset( self::$stylesheet_slug )
```
but rather this:
```php
			! (
				isset( self::$template_directory )
				&&
				isset( self::$template_slug )
				&&
				isset( self::$stylesheet_directory )
				&&
				isset( self::$stylesheet_slug )
			)
```
which is equivalent to your original code.

Reply:
Yes, it confuses me. 

I've simplified it by removing the conditionals entirely in 64c84073a.